example/settings.txt
CLIENT_REQUEST_LATENCY = 10
CLIENT_NETWORK_STRETCH = 1
MIN_DELAY = 5
MAX_DELAY = 10
VALUE_SIZE = 128

CICLES_TO_ONE_MILLI_SECOND = 100

BANDWIDTH = 125000000

CLIENTS_PER_DATACENTER = 200

REST_TIME = 500
REST_TIME_INTERVAL = 50
CLIENT_READ_PERCENTAGE = 90
CLIENT_UPDATE_PERCENTAGE = 10

CLIENT_EAGER_PERCENTAGE = 0



TOTAL_OBJECTS_PER_DATACENTER = 1000

LEVELS_PERCENTAGE = [50,25,15,10]

STATISTICS_WINDOW = 1

SHOULD_PARTITION_DC = true
SHOULD_PARTITION_CLIENTS = false

DC_PARTITION_LEVEL = 1
CLIENTS_PARTITION_LEVEL = 1


PARTITION_START_PERCENTAGE = 50
PARTITION_STOP_PERCENTAGE = 70

#Luis Configs
PARTITIONS_ARE_DELAYS = true
PARTITION_STRETCH_L1_PERCENTAGE = 800
PARTITION_STRETCH_L2_PERCENTAGE = 800
PARTITION_STRETCH_L3_PERCENTAGE = 800
PARTITION_MESSAGE_L1_AFFECTED_PERCENTAGE = 100
PARTITION_MESSAGE_L2_AFFECTED_PERCENTAGE = 100
PARTITION_MESSAGE_L3_AFFECTED_PERCENTAGE = 100

TOPOLOGY_FILE = example/topologies/eight_nodes_simple.top

# options: "throughput" or "accept"
METRICS = throughput

example/config-cops.txt
# network size
SIZE 8

# parameters of periodic execution
CYCLES 100000
CYCLE 1

# random.seed 1234567890
network.size SIZE
simulation.endtime CYCLES
simulation.logtime CYCLE

groups {[0],[0,1],[0,1,2,3],[0,1,2,3,4,5,6,7]}{[1],[0,1],[0,1,2,3],[0,1,2,3,4,5,6,7]}{[2],[2,3],[0,1,2,3],[0,1,2,3,4,5,6,7]}{[3],[2,3],[0,1,2,3],[0,1,2,3,4,5,6,7]}{[4],[4,5],[4,5,6,7],[0,1,2,3,4,5,6,7]}{[5],[4,5],[4,5,6,7],[0,1,2,3,4,5,6,7]}{[6],[6,7],[4,5,6,7],[0,1,2,3,4,5,6,7]}{[7],[6,7],[4,5,6,7],[0,1,2,3,4,5,6,7]}














transport urt
tree tree
##########################################

################### protocols ===========================

protocol.link peersim.core.IdleProtocol

protocol.tree example.cops.TreeProtocol
protocol.tree.step CYCLE
protocol.tree.transport tr
protocol.tree.tree_protocol tree
protocol.tree.linkable link


protocol.urt example.common.PointToPointTransport

protocol.tr UnreliableTransport
protocol.tr.transport urt
protocol.tr.drop 0

################### initialization ======================

### Creates graph and sets latencies
init.1 example.common.WireTopology
init.1.protocol link

init.2 example.cops.InitTreeProtocol
init.2.tree_protocol tree
init.2.clients_cycle 1

init.sch CDScheduler
init.sch.protocol tree
init.sch.randstart

################ control ==============================

control.0 example.cops.Controller
control.0.protocol tree
control.0.step CYCLE
control.0.output_file output/

CHANGELOG
MJ 07/07/09
- When there was a loop in the filesystem under one of the directories in
  the classpath, peersim hung when trying to collect the available classes
  (in order to resolve the underspecified classnames in the config). I added
  a temporary fix, which stops the directory walk after a (very large) maximal
  number of visited directories. Now peersim does not hang, but either fails
  with a clear message, or executes the simulation. Note that jar files
  are not affected, and they can still be arbitrarily large. When one needs
  an excessive number of packages (thousands) then a jar file is a better
  idea.
- Added a new constructor to RandPermutation for convenience.

MJ 09/06/09
- Fixed a bug in Scheduler found by Alberto. In the event
  based engine if "from" was set, the component was
  executed at "from" irrespective of "until". Besides,
  a negative "step" was not checked for.

AM 07/06/09
- Solved a bug identified by Mark: constant expressions
  were not allowed on RangeSimulator ranges, although they
  were explicitly mentioned in the documentation. The doc
  has been improved as well, saying that variables are not
  allowed inside ranges.
- Changed constants in peersim.rangesime.TaggedOutputStream
  from protected to public; this is needed to use the same
  class in other simulator implementations. Added the @config
  specification, so they appear correctly in the documentation.

AM 31/05/09
- Solved a bug in peersim.edsim.EDSimulator
  ClassCastException generated inside a user Control object
  were caught by the EDSimulator and a misleading error
  message was printed.

AM 30/04/09
- Solved a bug in ConfigContainer.  
  ConfigContainer.getInstance(String name, Object def) was 
  returning the default object even if name was defined but
  a parameter was missing in the definition of the created
  object. Thanks to Marco Biazzini for spotting this.
- Improved the functionality of configuration file parsing
  in peersim.config.ConfigProperties. Now, each of the
  command line parameter is first interpreted as a file
  name, and if this file does not exist, it is set as
  a configuration property. This allows the configuration 
  to be split in multiple files.
  Thanks to Jordi Pujol for suggesting the idea and
  implementing a first version of it.  

AM 8/11/2008
VERSION 1.0.4 RELEASED
------------------------------------------------------------------
------------------------------------------------------------------

AM 8/10/08
- Modified EDSimulator to print additional information in case
  of protocols not implementing a particular interface. Class cast
  exceptions launched by event processing are not swallowed any
  more.

AM 6/6/08
- Solved a bug in EDSimulator. Time was not reset to zero when network
  is re-initialized, this caused some problems in case of multiple
  experiments.

AM 24/05/08
- Solved a bug in ConfigContainer; some exceptions were incorrectly
  swallowed.

AM 23/05/08
- Improved KingParser; it now performs some check about file format, and
  print out the appropriate exceptions.

MJ 21/05/08
- Fixed serious initialization bug in MedianStats
- Changed OracleIdleProtocol, UniformRandomTransport and UnreliableTransport
  into final, due to their unconventional implementation of "clone" (they are
  singletons)
- Fixed minor issues in IncrementalFreq related to java conventions on equals
  and hashCode
- Added "close" for various unclosed streams (no serious issues, but it was
  not nice in any case)
- Various minor issues found by FindBugs (that is also responsible for the
  items above)

MJ 20/04/08
- Refactoring of EDSimulator to support the application of any event queue
  implementation. The even queue can now be configured to be any extrenal class
  implementing peersim.edsim.PriorityQ. Tests show there is no noticeable
  penalty for performance. (credit to Jan Schoenherr for pointing out some
  issues with edsim)
- Fixing various problems with handling integer overflow in time management in
  EDSimulator, Heap, and Scheduler.

MJ 18/04/08
- Changed truncation to rounding (as stated in the class docs) for
  DynamicNetwork. This fixes issues with small network sizes (or small "add"
  percentages).
- Fixed bug in RandNI: it assumed it is called after insertion, while the
  specification says it is called before insertion. This caused misbehavior for
  networks of size 1, and, more annoyingly, resulted in the node with index 0
  _never_ being linked to in any network of any size
  (note that for simulations that use shuffle (and most do)
  this is not noticable) (credit for spotting goes to Balazs Banhelyi).

AM 07/04/08
- Applied the javadoc patch of Filippo Giunchedi

MJ 18/03/08
- Added a "k" parameter to WireFromFile
- Added some popular file formats to GraphPrinter (as well as GraphIO)

MJ 24/02/08
- Added a "background" parameter to PeakDistribution and improved the docs of
  DynamicNetwork and OscillatingNetwork.

AM 04/02/08
- Solved a bug in phase management for both cd and ed
  simulations: at the end of the first experiment, phase was set
  to POST_SIMULATION and in subsequent experiments it stayed that way
  (never set back to UNKNOWN_PHASE).
  (Credits to Alessandro Russo)

MJ 19/01/08
- Changed default of parameter "maxsize" in DynamicNetwork and
  OscillatingNetwork. Now the default is: no limit on the size.
- Changed confusing parameter name "network.maxSize" to
  "network.initialCapacity".
- Fixed a bug in Network.setCapacity: now it removes nodes correctly.

AM 24/12/2007
VERSION 1.0.3 RELEASED
------------------------------------------------------------------
------------------------------------------------------------------

AM 18/12/07
- Reverted peersim.core.Scheduler to 1.19 version (removing Spyros
  changes), but adding the control to avoid at and step/from/until
  at the same time.

MJ 08/12/07
- Fixed RandRemoval: now it accepts parameter "n=1" (credits to Norbert
  Tolgyesi for spotting this)

MJ 04/11/07

src/example/common/Settings.java
package example.common;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

public class Settings {

    private Settings() {

    }

    private static Properties prop = new Properties();
    static {
        try (InputStream input = new FileInputStream("example/settings.txt")) {
            // load a properties file
            prop.load(input);
        } catch (IOException ex) {
            ex.printStackTrace();
            System.exit(-2);
        }
    }


    public static final int CLIENT_REQUEST_LATENCY = Integer.parseInt(prop.getProperty("CLIENT_REQUEST_LATENCY"));
    public static final float CLIENT_NETWORK_STRETCH = Float.parseFloat(prop.getProperty("CLIENT_NETWORK_STRETCH"));
    public static final int MIN_DELAY = Integer.parseInt(prop.getProperty("MIN_DELAY"));
    public static final int MAX_DELAY = Integer.parseInt(prop.getProperty("MAX_DELAY"));

    public static final int VALUE_SIZE = Integer.parseInt(prop.getProperty("VALUE_SIZE"));

    public static final int BANDWIDTH = Integer.parseInt(prop.getProperty("BANDWIDTH"));

    public static final int CICLES_TO_ONE_MILLI_SECOND= Integer.parseInt(prop.getProperty("CICLES_TO_ONE_MILLI_SECOND"));



    // How long it takes for a client to connect to  other DC
    // CLIENT SETTINGS

    public static final int CLIENTS_PER_DATACENTER = Integer.parseInt(prop.getProperty("CLIENTS_PER_DATACENTER"));

    // REST_TIME = intervalo entre pedidos
    // REST_TIME_INTERVAL = variancia do intervalo

    public static final int REST_TIME = Integer.parseInt(prop.getProperty("REST_TIME"));
    public static final int REST_TIME_INTERVAL = Integer.parseInt(prop.getProperty("REST_TIME_INTERVAL"));
    public static final int CLIENT_READ_PERCENTAGE = Integer.parseInt(prop.getProperty("CLIENT_READ_PERCENTAGE"));
    public static final int CLIENT_UPDATE_PERCENTAGE = Integer.parseInt(prop.getProperty("CLIENT_UPDATE_PERCENTAGE"));

    // percentagem de clientes que nao usa restime
    public static final int CLIENT_EAGER_PERCENTAGE = Integer.parseInt(prop.getProperty("CLIENT_EAGER_PERCENTAGE"));

    public static final String CLIENT_MIGRATION_ODDS = prop.getProperty("CLIENT_MIGRATION_ODDS");

    public static final String CLIENT_OBJECT_READ_LVL_0 = prop.getProperty("CLIENT_OBJECT_READ_LVL_0");
    public static final String CLIENT_OBJECT_READ_LVL_1 = prop.getProperty("CLIENT_OBJECT_READ_LVL_1");
    public static final String CLIENT_OBJECT_READ_LVL_2 = prop.getProperty("CLIENT_OBJECT_READ_LVL_2");
    public static final String CLIENT_OBJECT_READ_LVL_3 = prop.getProperty("CLIENT_OBJECT_READ_LVL_3");

    public static final String CLIENT_OBJECT_UPDATE_LVL_0 = prop.getProperty("CLIENT_OBJECT_UPDATE_LVL_0");
    public static final String CLIENT_OBJECT_UPDATE_LVL_1 = prop.getProperty("CLIENT_OBJECT_UPDATE_LVL_1");
    public static final String CLIENT_OBJECT_UPDATE_LVL_2 = prop.getProperty("CLIENT_OBJECT_UPDATE_LVL_2");
    public static final String CLIENT_OBJECT_UPDATE_LVL_3 = prop.getProperty("CLIENT_OBJECT_UPDATE_LVL_3");

    public static final int TOTAL_OBJECTS_PER_DATACENTER = 1000;

    // Distribuicao dos objectos em cada datacenter
    public static final String LEVELS_PERCENTAGE = prop.getProperty("LEVELS_PERCENTAGE");

    public static final float STATISTICS_WINDOW = Float.parseFloat(prop.getProperty("STATISTICS_WINDOW"));

    public static final boolean SHOULD_PARTITION_DC = Boolean.parseBoolean(prop.getProperty("SHOULD_PARTITION_DC"));
    public static final boolean SHOULD_PARTITION_CLIENTS = Boolean.parseBoolean(prop.getProperty("SHOULD_PARTITION_CLIENTS"));

    // Partition lvl can be 1, 2, or 3. Any other value will load the file
    // eight_nodes_partition_custom (for projects without tree)
    // or eight_nodes_seven_brokers_partition_custom (for projects with tree)
    // Partition lvl 1: Servers cannot comunicate with other servers
    // Partition lvl 2: Each server can communicate with 1 server
    // Partition lvl 3: Each server can communicate with 3 servers
    public static final int DC_PARTITION_LEVEL = Integer.parseInt(prop.getProperty("DC_PARTITION_LEVEL"));
    public static final int CLIENTS_PARTITION_LEVEL = Integer.parseInt(prop.getProperty("CLIENTS_PARTITION_LEVEL"));


    public static final double PARTITION_START_PERCENTAGE = Double.parseDouble(prop.getProperty("PARTITION_START_PERCENTAGE"));
    public static final double PARTITION_STOP_PERCENTAGE = Double.parseDouble(prop.getProperty("PARTITION_STOP_PERCENTAGE"));

    // Luis Configs
    public static final boolean PARTITIONS_ARE_DELAYS = Boolean.parseBoolean(prop.getProperty("PARTITIONS_ARE_DELAYS"));
    public static final double PARTITION_STRETCH_L1_PERCENTAGE = Double.parseDouble(prop.getProperty("PARTITION_STRETCH_L1_PERCENTAGE"));
    public static final double PARTITION_STRETCH_L2_PERCENTAGE = Double.parseDouble(prop.getProperty("PARTITION_STRETCH_L2_PERCENTAGE"));
    public static final double PARTITION_STRETCH_L3_PERCENTAGE = Double.parseDouble(prop.getProperty("PARTITION_STRETCH_L3_PERCENTAGE"));
    public static final long PARTITION_MESSAGE_L1_AFFECTED_PERCENTAGE = Long.parseLong(prop.getProperty("PARTITION_MESSAGE_L1_AFFECTED_PERCENTAGE"));
    public static final long PARTITION_MESSAGE_L2_AFFECTED_PERCENTAGE = Long.parseLong(prop.getProperty("PARTITION_MESSAGE_L2_AFFECTED_PERCENTAGE"));
    public static final long PARTITION_MESSAGE_L3_AFFECTED_PERCENTAGE = Long.parseLong(prop.getProperty("PARTITION_MESSAGE_L3_AFFECTED_PERCENTAGE"));


    public static final String TOPOLOGY_FILE = prop.getProperty("TOPOLOGY_FILE");
    public static final String METRICS = prop.getProperty("METRICS");

    // DEBUG
    public static final boolean PRINT_INFO = true;
    public static final boolean PRINT_CYCLE = false;

    public static final boolean PRINT_VERBOSE = true;
    public static final String CLIENT_LOCALITY_PERCENTAGE = "[0,0,0,100]";
    public static final String CLIENT_READ_LEVEL_PERCENTAGE = "[50,25,15,10]";
    public static final String CLIENT_UPDATE_LEVEL_PERCENTAGE = "[50,25,15,10]";
}
